compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala

```diff
@@ -334,29 +375,42 @@ abstract class PrepJSInterop extends plugins.PluginComponent
           reporter.error(implDef.pos, s"${sym.nameString} extends ${badName} " +
               "which does not extend js.Any.")
 
-        // Check that we are not an anonymous class
+        // Check that @ScalaJSDefined is not used on an object
+        case _ if !isJSNative && !sym.isClass =>
+          reporter.error(implDef.pos,
+              "Objects cannot be Scala.js-defined")
+
+        // Check that a non-native JS class does not inherit directly from AnyRef
+        case _ if !isJSNative && !sym.isTrait &&
+            sym.info.parents.exists(_ =:= AnyRefClass.tpe) =>
+          reporter.error(implDef.pos,
+              "A Scala.js-defined JS class cannot extend directly AnyRef. " +
+              "It must extend a JS class (native or not).")
+
+        // Check that we do not inherit directly from a native JS trait
+        case _ if !isJSNative && sym.info.parents.exists(isNativeJSTraitType) =>
+          val obj = if (sym.isTrait) "trait" else "class"
+          reporter.error(implDef.pos,
+              s"A Scala.js-defined JS $obj cannot extend directly a native " +
+              "JS trait. It must extend a JS class (native or not).")
```

Comment:
The last sentence is misleading: This error condition can occur even if we directly extend a JS class (but also extend a native trait).

Also, a non-native trait can extend a non-native JS trait, can't it?